### PR TITLE
Stop using globally exposed DSL

### PR DESCRIPTION
**Pull request by pirj**
~I find it puzzling to explain why DSL was exposed~

This PR is a prerequisite for:
 - https://github.com/rspec/rspec-expectations/pull/1245
 - https://github.com/rspec/rspec-core/pull/2803
 - https://github.com/rspec/rspec-mocks/pull/1365

### CODE REVIEW

File: spec/support/shared_examples.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 require 'pathname'
 
-shared_examples_for "an rspec-rails example group mixin" do |type, *paths|
+RSpec.shared_examples_for "an rspec-rails example group mixin" do |type, *paths|
```
Comment (pirj) [inline spec/support/shared_examples.rb:3]: It seems that we require spec support files before `RSpec.configure`, and `Core::DSL.expose_globally!` was taking precedence. Interesting.

Comment (JonRowe) [reply spec/support/shared_examples.rb:3]: Sounds about right

**Review by JonRowe — APPROVED** [review]

**Comment by JonRowe** [comment]
This is fine, but do you mind not merging until I've released a release for 6.1?

**Comment by JonRowe** [comment]
(As RSpec 4 will be sometime in the new year I suspect)

**Comment by pirj** [comment]
I had to make yet another change here, check If `RSpec::Matchers.configuration.respond_to?(:syntax)` to support both RSpec 3 & 4.

@JonRowe Do you mind if I merge this before we release `rspec-rails` 4.1? There are still a couple of issues to address there.

The usage of globally exposed DSL and removed `disable_monkey_patching!`/`configuration.syntax` forces me to hack zero monkey patching PRs to point to this PR (and all subsequent PRs that I intend to send for RSpec 4.0):
```diff
-  clone_repo "rspec-rails" "main"
+  clone_repo "rspec-rails" "remove-monkey-patching"
```

**Comment by pirj** [comment]
Thanks!

**Pull request merged**